# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/GraphDriver.cpp
# Commit: f81e1b1615b3
# Full Hash: f81e1b1615b3a9c5ef6c62a291f79029b3807dde
# Author: Karl Tomlinson <karlt+@karlt.net>
# Date: 2023-03-31 09:51:00
# Regressor Bug: 1801190
# File Overlap Count: 1
# Description:
#   Bug 1801190 do not reset mAudioStreamState to None on state changes from an old cubeb_stream r=pehrsons
#   
#   Such resets were preventing transitions to Running for new streams and allowing
#   MaybeStartAudioStream() to start another stream.
#   
# ==============================================================================

diff -r 9592d9865f5a -r f81e1b1615b3 dom/media/GraphDriver.cpp
--- a/dom/media/GraphDriver.cpp	Fri Mar 31 02:23:14 2023 +0300
+++ b/dom/media/GraphDriver.cpp	Thu Mar 30 23:18:44 2023 +0000
@@ -1048,12 +1048,21 @@
   LOG(LogLevel::Debug,
       ("AudioCallbackDriver(%p) State: %s", this, StateToString(aState)));
 
-  AudioStreamState streamState = mAudioStreamState;
-  if (aState != CUBEB_STATE_STARTED) {
-    // Clear the flag for the not running states: stopped, drained, error.
-    streamState = mAudioStreamState.exchange(AudioStreamState::None);
+  if (aState == CUBEB_STATE_STARTED) {
+    return;
   }
 
+  if (!IsStarted()) {
+    // mAudioStream has already entered STOPPED, DRAINED, or ERROR.
+    // Don't reset a Pending state indicating that a task to destroy
+    // mAudioStream and init a new cubeb_stream has already been triggered.
+    return;
+  }
+
+  // Reset for the not running states: stopped, drained, error.
+  AudioStreamState streamState =
+      mAudioStreamState.exchange(AudioStreamState::None);
+
   if (aState == CUBEB_STATE_ERROR) {
     // About to hand over control of the graph.  Do not start a new driver if
     // StateCallback() receives an error for this stream while the main thread